testutil/network/basenet/network.go

```diff
@@ -0,0 +1,125 @@
+package basenet
+
+import (
+	"context"
+	"sync/atomic"
+	"testing"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	sdknetwork "github.com/cosmos/cosmos-sdk/testutil/network"
+	"github.com/stretchr/testify/require"
+
+	"github.com/pokt-network/poktroll/testutil/network"
+	"github.com/pokt-network/poktroll/testutil/testkeyring"
+)
+
+var _ network.InMemoryNetwork = (*BaseInMemoryNetwork)(nil)
+
+// BaseInMemoryNetwork is an "abstract" (i.e. partial) implementation, intended
+// to be embedded by other ("concrete") InMemoryNetwork implementations.
+type BaseInMemoryNetwork struct {
+	Config                      network.InMemoryNetworkConfig
+	PreGeneratedAccountIterator *testkeyring.PreGeneratedAccountIterator
+	Network                     *sdknetwork.Network
+
+	// lastAccountSeqNumber stores the last (most recently generated) account sequence number.
+	// NB: explicitly NOT using atomic.Int32 as it's usage doesn't compose well with anonymous
+	// literal declarations.
+	lastAccountSeqNumber int32
+}
+
+// NewBaseInMemoryNetwork creates a new BaseInMemoryNetwork with the given
+// configuration and pre-generated accounts. Intended to be used in constructor
+// functions of structs that embed BaseInMemoryNetwork.
+func NewBaseInMemoryNetwork(
+	t *testing.T,
+	cfg *network.InMemoryNetworkConfig,
+	preGeneratedAccounts *testkeyring.PreGeneratedAccountIterator,
+) *BaseInMemoryNetwork {
+	t.Helper()
+
+	return &BaseInMemoryNetwork{
+		Config:                      *cfg,
+		PreGeneratedAccountIterator: preGeneratedAccounts,
+
+		// First functional account sequence number is 1. Starting at 0 so that
+		// callers can always use NextAccountSequenceNumber() (no boundary condition).
+		lastAccountSeqNumber: int32(0),
+	}
+}
+
+// InitializeDefaults sets the underlying cosmos-sdk testutil network config to
+// a reasonable default in case one was not provided with the InMemoryNetworkConfig.
+func (memnet *BaseInMemoryNetwork) InitializeDefaults(t *testing.T) {
+	if memnet.Config.CosmosCfg == nil {
+		t.Log("Cosmos config not initialized, using default config")
+
+		// Initialize a network config.
+		cfg := network.DefaultConfig()
+		memnet.Config.CosmosCfg = &cfg
+	} else {
+		t.Log("Cosmos config already initialized, using existing config")
+	}
+}
+
+// GetClientCtx returns the underlying cosmos-sdk testutil network's client context.
+func (memnet *BaseInMemoryNetwork) GetClientCtx(t *testing.T) client.Context {
+	t.Helper()
+
+	// Only the first validator's client context is populated.
+	// (see: https://pkg.go.dev/github.com/cosmos/cosmos-sdk/testutil/network#pkg-overview)
+	ctx := memnet.GetNetwork(t).Validators[0].ClientCtx
+
+	// TODO_UPNEXT(@bryanchriswhite): Ensure validator key is always available.
```

Comment:
```suggestion
	// TODO_UPNEXT(@bryanchriswhite): Ensure validator key is always available.
	ctx := memnet.GetNetwork(t).Validators[0].ClientCtx
```

Reply:
This is more related to the line below it that the one above. I didn't want to crowd either comment.